### PR TITLE
perf(ingester): deferred sort key load

### PR DESCRIPTION
**Pull request by domodwyer**
This PR adds a mechanism to load the sort key from the catalog on demand, or in the background after a set period of time. This is instead of at construction time, or `persist()` time, as large numbers of instances are both constructed (in the hot path) and `persist()` at the same time, causing a large spike in catalog load (#5767).

The exact timing of the background load is uniformly randomly for each instance - effectively a random jitter. This should smear the queries across the time window, and reduce the peak catalog load.

## Why Not Include in `PartitionCache`

At the time of writing there are 1,249,512 partitions, with an average sort key length of 37 bytes (as string content, ignoring overhead of heap pointers, etc) and a maximum of **58818** bytes. Caching these is prohibitively expensive, and this deferred loading construct will prove useful and be reused in a future PR to load other strings needed at persist time.

## New Invariant

As this is effectively a lazy-loaded cache, correctness is violated if there are multiple writers and no cache invalidation.

Currently today only the ingester **writes** to the sort key in the catalog - as long as this is maintained, this implementation will be correct.

NOTE: this is currently loaded, but unused - actually using it is blocked until finer-grained locking can be implemented across the partition (this is what I actively working on - relates to #5725).

---

* feat: deferred partition sort key fetcher (c022ab678)

      Adds a new DeferredSortKey type that fetches a partition's sort key from
      the catalog in the background, or on-demand if not yet pre-fetched.
      
      From the caller's perspective, little has changed compared to reading it
      from the catalog directly - the sort key is always returned when calling
      get(), regardless of the mechanism, and retries are handled
      transparently. Internally the sort key MAY have been pre-fetched in the
      background between the DeferredSortKey being initialised, and the call
      to get().
      
      The background task waits a (uniformly) random duration of time before
      issuing the catalog query to pre-fetch the sort key. This allows large
      numbers of DeferredSortKey to (randomly) smear the lookup queries over a
      large duration of time. This allows a large number of DeferredSortKey to
      be initialised in a short period of time, without creating an equally
      large spike in queries against the catalog in the same time period.

* perf(ingester): deferred sort key lookup queries (afcb96ae4)

      This commit carries the SortKey in the PartitionData, and configures the
      ingester to use deferred sort key lookups, smearing the lookups across a
      fixed period of time after initialising the PartitionData, instead of
      querying for the sort key at persist time.
      
      This allows large numbers of PartitionData to be initialised without
      causing a equally large spike in catalog load to resolve the sort key -
      instead this load is spread out randomly to reduce peak query rps.

### CODE REVIEW

File: ingester/src/data/partition/resolver/cache.rs [hunk 1]
```diff
@@ -43,6 +45,18 @@ struct Entry {
 /// Each cache hit _removes_ the entry from the cache - this eliminates the
 /// memory overhead for items that were hit. This is the expected (only valid!)
 /// usage pattern.
+///
+/// # Deferred Sort Key Loading
+///
+/// This cache does NOT cache the [`SortKey`] for each [`PartitionData`], as the
+/// sort key can be large and is likely unique per table, and thus not
```
Comment (NGA-TRAN) [inline ingester/src/data/partition/resolver/cache.rs:52]:
```suggestion
/// sort key can be large and is likely unique per partition, and thus not
```

**Review by NGA-TRAN — APPROVED** [review]
The logic that delay reading sort key sounds good to me.

File: ingester/src/data/partition/resolver/sort_key.rs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+/// This implementation combines lazy / deferred loading of the [`SortKey`] from
+/// the [`Catalog`], and a background timer that pre-fetches the [`SortKey`]
+/// after some random duration of time. Combined, these behaviours smear the
+/// [`SortKey`] queries across the allowable time range, avoiding a large number
+/// of queries from executing when multiple [`SortKey`] are needed in the system
+/// at one point in time.
+///
+/// If the [`DeferredSortKey`] is dropped and the background task is still
+/// incomplete (sleeping / actively fetching the [`SortKey`]) it is aborted
+/// immediately. The background task exists once it has successfully fetched the
```
Comment (carols10cents) [inline ingester/src/data/partition/resolver/sort_key.rs:33]: I *think* this is supposed to say:

```suggestion
/// immediately. The background task exits once it has successfully fetched the
```

**Review by carols10cents — DISMISSED** [review]
Tiny things but I think should be fixed :)

File: ingester/src/data/partition/resolver/sort_key.rs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+            // If there is a resolved value, return it.
+            if let State::Resolved(v) = &*state {
+                return v.clone();
+            }
+        }
+
+        // Otherwise resolve the value immediately, aborting the background
+        // task.
+        let sort_key = fetch(self.partition_id, &*self.catalog, &self.backoff_config).await;
```
[the remark this answers is not in the document]
Comment (domodwyer) [reply ingester/src/data/partition/resolver/sort_key.rs:130]: I wrote it like this to make this fn abort-safe - killing the background task before polling `fetch()` could leave us with no background fetch if the future at the `fetch().await` was dropped/not completed.

But looking at this again, it doesn't really matter - a caller who needs the sort key will eventually call this and wait long enough to load it, even if the background task was aborted.

There's also a further subtlety - multiple concurrent calls to this method will stack up and all try to fetch the value. I have documented this in ab78f99ab2e3de07d210d115135b10720fdec0d2 too - it think this is "good enough" for now.

File: ingester/src/data/partition/resolver/sort_key.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+            .await
+            .partitions()
+            .update_sort_key(partition_id, &["uno", "dos", "bananas"])
+            .await
+            .expect("should update existing partition key");
+
+        // Read the updated sort key
+        let fetched = DeferredSortKey::new(
+            partition_id,
+            Duration::from_secs(10_000),
```
[the remark this answers is not in the document]
Comment (domodwyer) [reply ingester/src/data/partition/resolver/sort_key.rs]: 4518bd49d1df0ccdda4825ed991bcb0660f10b2e

**Pull request merged**